File: libs/statemanager/src/main/kotlin/mono/state/FileMediator.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state

import kotlinx.browser.document
import mono.html.A
import mono.html.Input
import mono.html.InputType
import mono.html.setAttributes
import org.w3c.dom.url.URL
import org.w3c.files.Blob
import org.w3c.files.FileList
import org.w3c.files.FileReader
import org.w3c.files.get

/**
 * A mediator class for file interactions.
 */
internal class FileMediator {
    fun saveFile(jsonString: String) {
        document.body?.run {
            val fileBlob = Blob(arrayOf(jsonString))
            val node = A(classes = "hidden") {
                href = URL.Companion.createObjectURL(fileBlob)
                setAttributes("download" to "$DEFAULT_FILENAME.$EXTENSION")
            }
            node.click()
            node.remove()
        }
    }

    fun openFile(onFileLoadedAction: (String) -> Unit) {
        document.body?.run {
            val fileInput = Input(InputType.FILE, classes = "hidden") {
                setAttributes("accept" to ".$EXTENSION")
            }
            fileInput.onchange = {
                readFile(fileInput.files, onFileLoadedAction)
                fileInput.remove()
            }
            fileInput.click()
        }
    }

    private fun readFile(fileList: FileList?, onFileLoadedAction: (String) -> Unit) {
        val selectedFile = fileList?.get(0) ?: return
        val reader = FileReader()
        reader.onload = {
            val text = reader.result.toString()
            onFileLoadedAction(text)
        }
        reader.readAsText(selectedFile)
    }

    companion object {
        private const val DEFAULT_FILENAME = "monosketch"
        private const val EXTENSION = "mono"
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/onetimeaction/file-mediator.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

/**
 * A mediator class for file interactions.
 */
export class FileMediator {
    saveFile(filename: string, jsonString: string) {
        const fileBlob = new Blob([jsonString]);
        const node = document.createElement('a');
        node.href = URL.createObjectURL(fileBlob);
        node.setAttribute('download', `${filename}.${FileMediator.EXTENSION}`);
        node.style.display = 'none';
        document.body.appendChild(node);
        node.click();
        document.body.removeChild(node);
    }

    openFile(onFileLoadedAction: (content: string) => void) {
        const fileInput = document.createElement('input');
        fileInput.type = 'file';
        fileInput.accept = `.${FileMediator.EXTENSION}`;
        fileInput.style.display = 'none';
        fileInput.onchange = () => {
            this.readFile(fileInput.files, onFileLoadedAction);
            document.body.removeChild(fileInput);
        };
        document.body.appendChild(fileInput);
        fileInput.click();
    }

    private readFile(fileList: FileList | null, onFileLoadedAction: (content: string) => void) {
        const selectedFile = fileList?.[0];
        if (!selectedFile) return;

        const reader = new FileReader();
        reader.onload = () => {
            const text = reader.result as string;
            onFileLoadedAction(text);
        };
        reader.readAsText(selectedFile);
    }

    private static readonly EXTENSION = 'mono';
}
